### PR TITLE
Added more SpriteKit extensions

### PR DESCRIPTION
**Pull request by moyerr**
Helping out with Issue #303.

## Checklist

- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 4.
- [x] New extensions support iOS 8.0+ / tvOS 9.0+ / macOS 10.10+ / watchOS 2.0+.
- [ ] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.

### CODE REVIEW

File: Sources/Extensions/SpriteKit/SKPhysicsBodyExtensions.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    ///
+    ///     if player.physicsBody?.isTouching(Category.fire) {
+    ///         // Reduce player health
+    ///     }
+    ///
+    /// - Parameter category: The `categoryBitmMask` to check
+    /// - Returns: `true` if the physics body is contacting another body
+    ///             with the given bitmask, `false` otherwise.
+    public func isTouching(_ category: UInt32) -> Bool {
+        return allContactedBodies().contains(where: { $0.categoryBitMask == category })
```
Comment (guykogus) [inline Sources/Extensions/SpriteKit/SKPhysicsBodyExtensions.swift]: This is incorrect. To check that a bitmask intersects you need use a logical AND:
`($0.categoryBitMask & category) != 0`.

File: Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#if canImport(SpriteKit)
+import SpriteKit
+
+// MARK: - Properties
+public extension SKPhysicsContact {
+    
+    /// SwifterSwift: A convenience variable for accessing the
+    /// nodes involved in a physical contact.
+    public var nodes: [SKNode] {
+        return [bodyA.node, bodyB.node].compactMap({ $0 })
```
Comment (guykogus) [inline Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift]: You can remove the parentheses.
```Swift
return [bodyA.node, bodyB.node].compactMap { $0 }
```

File: Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    ///
+    /// - Parameters:
+    ///     - categoryA: The first category to check
+    ///     - categoryB: The second category to check
+    ///
+    /// - Returns: `true` if the collision involves the two specific
+    ///     categories, `false` otherwise.
+    ///
+    public func involves(_ categoryA: UInt32, and categoryB: UInt32) -> Bool {
+        return (bodyA.categoryBitMask == categoryA && bodyB.categoryBitMask == categoryB) ||
```
Comment (guykogus) [inline Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift]: Again, you need to use logical AND.

File: Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    ///     }
+    ///
+    /// - Parameters:
+    ///     - categoryA: The first category to check
+    ///     - categoryB: The second category to check
+    ///
+    /// - Returns: `true` if the collision involves the two specific
+    ///     categories, `false` otherwise.
+    ///
+    public func involves(_ categoryA: UInt32, and categoryB: UInt32) -> Bool {
```
Comment (guykogus) [inline Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift]: I feel like `contains` would be a better name, so it's used quite often in Swift.

File: Sources/Extensions/SpriteKit/SKSpriteNodeExtensions.swift [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        timePerFrame seconds: TimeInterval,
+        resize: Bool,
+        restore: Bool,
+        completion: (() -> Void)? = nil) {
+        
+        let textureAtlas = SKTextureAtlas(named: atlasName)
+        
+        let textures = textureAtlas.textureNames
+            .sorted()
+            .map({ SKTexture(imageNamed: $0) })
```
Comment (guykogus) [inline Sources/Extensions/SpriteKit/SKSpriteNodeExtensions.swift]: Again you can remove parentheses. I feel like we should be using SwiftLint to check for these :)
```Swift
.map { SKTexture(imageNamed: $0) }
```

File: Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift (hunk 4 above)
Comment (moyerr) [reply Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift]: This one I disagree with. `contains` is indeed used often in Swift, however, in this instance I think saying physical contact *involves* two object is more accurate than saying it *contains* two objects.

**Comment by moyerr** [comment]
@guykogus Any ideas for how to unit test this `SpriteKit` stuff?

**Comment by guykogus** [comment]
I guess you'll need to create a few `SKPhysicsBody` objects and place them in such a way that some should be touching while others are not and test your functions to see that they work.

As far as the animation function goes, maybe add an atlas with just a few frames (keep it as small as possible so as to not bloat the repo), run a test that runs the animation and then check it's state.
E.g.
```Swift
var isFinished = false
let timePerFrame = TimeInterval(1.0 / 30.0)
let spriteNode = SKSpriteNode()
spriteNode.animate(withAtlasNamed: "atlasName",
                   timePerFrame: timePerFrame,
                   resize: false,
                   restore: false,
                   completion: {
                    isFinished = true
})

XCTAssertFalse(isFinished)
Thread.sleep(forTimeInterval: timePerFrame)
XCTAssertFalse(isFinished)
Thread.sleep(forTimeInterval: timePerFrame)
XCTAssertTrue(isFinished)
```

File: Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift (hunk 4 above)
Comment (moyerr) [inline Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift]: @guykogus Would you accept the name as-is based on my previous comment?

**Comment by moyerr** [comment]
@guykogus Any idea why those tests fail on macOS but pass on iOS and tvOS?

File: Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift (hunk 4 above)
Comment (guykogus) [reply Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift]: Yeah, I think it's fine. I think I gave it a little 👍 before ;)

**Comment by guykogus** [comment]
No idea. The only thing I can think of is the timeout of the expectation being short, but that would surprise me. Have you tried running the macOS tests locally?

File: Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift (hunk 4 above)
Comment (moyerr) [reply Sources/Extensions/SpriteKit/SKPhysicsContactExtensions.swift]: @guykogus Thanks! It's still showing the ❌  by changes requested, which looks like it'll block merging.

**Comment by moyerr** [comment]
@guykogus Yep, it's saying:

> Asynchronous wait failed: Exceeded timeout

However, the expectation should be fulfilled immediately when the scene is presented, as it is in iOS and tvOS.

**Comment by SD10** [comment]
So I have no experience with `SpriteKit`, I'm kind of useless on this one

**Comment by guykogus** [comment]
It's been a long time since I used SpriteKit, and never used it on macOS, so can't really help you. Maybe try increase the timeout, but there may be something that it's expecting to happen...

**Comment by moyerr** [comment]
@guykogus So i just came to the realization that the `SKSpriteNode` animation method I added is probably useless, as the texture assets would be in a different Bundle from SwifterSwift. I think I should probably remove that one. 

Still not sure about the macOS unit tests failing.

**Comment by LucianoPAlmeida** [comment]
Closing this due inactivity.
Feel free to reopen it anytime :))

**Pull request closed without merging**